internal/resource-request-operator/broadcaster.go

```diff
@@ -0,0 +1,184 @@
+			updateResources(&currentResources, &oldNodeResources, &newNodeResources)
+			// node is starting, add all its resources.
+		} else if oldNode.Status.Phase == corev1.NodePending || oldNode.Status.Phase == corev1.NodeTerminated {
+			addResources(&currentResources, &newNodeResources)
+		}
+		// node is terminating or stopping, delete all its resources.
+	} else if oldNode.Status.Phase == corev1.NodeRunning &&
+		(newNode.Status.Phase == corev1.NodeTerminated || newNode.Status.Phase == corev1.NodePending) {
+		subResources(&currentResources, &oldNodeResources)
+	}
+	if err := b.writeResources(currentResources); err != nil {
+		klog.Errorf("OnUpdate error: unable to write allocatable of Node %s: %s", newNode.Name, err)
+	}
+}
+
+// react to a Node Delete.
+func (b *Broadcaster) onDelete(obj interface{}) {
+	node := obj.(*corev1.Node)
+	toDelete := &node.Status.Allocatable
+	currentResources := b.allocatable.DeepCopy()
+	if node.Status.Phase == corev1.NodeRunning {
+		subResources(&currentResources, toDelete)
+		if err := b.writeResources(currentResources); err != nil {
+			klog.Errorf("OnAdd error: unable to write allocatable of Node %s: %s", node.Name, err)
+		}
+	}
+}
+
+// write cluster resources in thread safe mode.
+func (b *Broadcaster) writeResources(newResources corev1.ResourceList) error {
+	b.offerMutex.Lock()
+	defer b.offerMutex.Unlock()
+	if newResources != nil {
+		b.allocatable = newResources.DeepCopy()
+		return nil
+	}
+
+	return fmt.Errorf("some error occurred during cluster resources read. Attempting writing nil resources")
+}
+
+// ReadResources return in thread safe mode a scaled value of the resources.
+func (b *Broadcaster) ReadResources() (corev1.ResourceList, error) {
+	b.offerMutex.RLock()
+	defer b.offerMutex.RUnlock()
+	if b.allocatable == nil {
+		return nil, fmt.Errorf("error getting cluster resources")
+	}
+	toRead := b.allocatable.DeepCopy()
+	for resourceName, quantity := range toRead {
+		scaled := quantity
+		b.scaleResources(&scaled)
+		toRead[resourceName] = scaled
+	}
+	return toRead, nil
+}
+
+func (b *Broadcaster) scaleResources(quantity *resource.Quantity) {
+	b.configMutex.RLock()
+	percentage := int64(b.clusterConfig.Spec.AdvertisementConfig.OutgoingConfig.ResourceSharingPercentage)
+	b.configMutex.RUnlock()
+	if percentage == 0 {
+		return
+	}
+
+	quantity.Set(quantity.Value() * percentage / 100)
+}
+
+// addResources is an utility function to add resources.
+func addResources(currentResources, toAdd *corev1.ResourceList) {
+	for resourceName, quantity := range *toAdd {
+		value := (*currentResources)[resourceName]
+		if !quantity.IsZero() {
+			value.Add(quantity)
+		}
+		(*currentResources)[resourceName] = value
+	}
+}
+
+// subResources is an utility function to subtract resources.
+func subResources(currentResources, toSub *corev1.ResourceList) {
+	for resourceName, quantity := range *toSub {
+		value := (*currentResources)[resourceName]
+		if !quantity.IsZero() {
```

Comment:
We check many times in the code that the quality has to be different from zero, is it a problem to add/subtract zero quantities?